### ISSUE TITLE
Vendoring from cisco?

### ISSUE THREAD

**Issue by hmaarrfk** [conda-forge/openh264-feedstock#5]
Does it makes sense to vendor this from Cisco's binaries so that we can make use them paying the patent fees?

https://github.com/conda-forge/ffmpeg-feedstock/issues/66

**Comment by sdvillal** [comment]
Can you find an example of a linux distro that does or cares about this?

**Comment by hmaarrfk** [comment]
mozilla cares about it. https://support.mozilla.org/en-US/kb/open-h264-plugin-firefox

An article about Fedora doing the same in 2016: https://www.phoronix.com/scan.php?page=news_item&px=Fedora-H264-OpenH264

**Comment by hmaarrfk** [comment]
Ubuntu doesn't distribute this: https://packages.ubuntu.com/search?keywords=openh264&searchon=names&suite=bionic&section=all presumably because 

Fedora has a wiki page explaining their logic: https://fedoraproject.org/wiki/OpenH264

**Comment by sdvillal** [comment]
In general people do not care much about openh264 because x264 is universally regarded as a more capable H264 encoder.

So let me rephrase the question. What can we make out of this listing: https://pkgs.org/download/x264? What I see is that usually the x264 packages are hosted in either community driven repositories (e.g. fedora or centos) or the official repos (e.g. arch or debian).

Are we worried about conda-forge being sued for patent royalties because of the content that a third party might create? If so, do we need to remove all traces of software capable of encoding (decoding?) content under a patent? And if so, do we need to work retroactively removing from our channels all versions of such software?

IANAL, but I do know the hardness we would endure if the answer to these questions would be "yes".

**Comment by hmaarrfk** [comment]
In my understanding, the consumer of the algorithm can be held liable for
the patent infringement.

Therefore, it is up to us, the consumers of our own packages to decide what
we want to be held accountable for infringing on these patents. As an
individual user, nobody will care, but as private companies, we might be
put in a situation where somebody audits the code we write and distribute.

So, I think as engineers that are using this library, it is safer to just
download it from Cisco since they provide a license for the encoder, so
long as it is downloaded by the end user of the application directly from
their site. The side-effect of using Cisco's binaries will not be slower
code, or hard to install (if we write the conda package right), but just a
more legally transparent move.


Those choosing to use x264 will have to weigh the risks themselves.

Hardware coders/decoder don't distribute the algorithm in their headers. I
presume that in by buying specific hardware, also purchasing a license to
use the codecs included.

Finally, engineers are free to use other codecs that aren’t incumbent on
any patents such as the VP series.

**Comment by sdvillal** [comment]
So you would propose this in order to give users an option to produce large amounts of H264 content without being held accountable for any patent royalties, not to protect CF, right? If that is the case and you manage to create that package with CISCO's h264 binary, then there is not much discussion this is an OK move entailing little technical burden. It also becomes a very low priority move.

So how would we proceed? Create a binary package + dev package, depend on dev when building but only on binary at runtime? What is th minimum C++ runtime needed by these precompiled binaries?

**Comment by hmaarrfk** [comment]
You are correct, protecting CF is not the motivation, rather protecting the users is.

I think the easiest is to make the conda package empty, and simply be a post-install script that fetches and downloads the binary from github. 

https://conda.io/docs/user-guide/tasks/build-packages/link-scripts.html

Apparently firefox does something to verify the download, I would have to look into that separately.

I think creating the dev package to save a few CI resources is overkill and might cause weird discrepancies.

**Comment by jakirkham** [comment]
This is an interesting line of discussion. Generally we have preferred building from source wherever possible. Though if there are some subtleties legally with that approach, this may require closer inspection. As conda-forge is NumFOCUS sponsored, we do have access to legal advice. Perhaps this is a good stage to get that feedback? A little unsure about next steps as this is not something I have lots of experience with. Thoughts on how best to proceed @conda-forge/core?

**Comment by sdvillal** [comment]
Downloading in a post hook is far from ideal because a) we add a second point of failure b) we get files out of conda control. I would go for the earliest of these that retain the patent protection:

- Mimic current package with headers, but binary vendored from cisco.
- Dev package with headers, runtime with binary vendored from cisco.
- The post hook thingy

In any case, we also need to package headers for downstream dependencies.

I personally think that if protecting the user is the goal, we should just not do anything that is too problematic from a maintenance perspective. If a user knows she might be liable for paying patents, she will know how to take care by herself. If she does not know, it is likely she won't use openh264 at all.

**Comment by hmaarrfk** [comment]
Unfortunately this video seems to detail that the download from the cisco distribution must happen by the end-user's computer.
http://vimeo.com/cullenfluffyjennings/openh264-ipr

It seems to say that even if Mozilla can create a binary signature, it can't actually send the compiled module within firefox itself.

By analogy, it seems that if we ask travis to to place the `so` from cisco in a `tar.gz` for conda and host it on anaconda.org cisco would not pay for the distribution fees. 

I would be most interested in hearing the lawyer's conclusions.

**Comment by scopatz** [comment]
@hmaarrfk @sdvillal - can someone type up exactly what questions (with full context and information) that you would like to ask our legal counsel.  That is probably the first step to getting an answer here.

**Comment by hmaarrfk** [comment]
Circling back to some of my older PRs

<details> <summary> email </summary>

# Legal implications of distributed h264 and h265 related software
March XX, 2019
conda-forge

It has come to our attention that certain comporate entities claim patents on some open source software distributed by conda-forge. 

Specifically, we are concerned about the implications to conda-forge and its community surrounding the h264 and h265 codecs that are claimed by MPEG-LA. We are unsure if at least two packages packages, x264 (GPL) and openh264 (BSD), could be considered as infringing on these patents.

In a [shared presentation](http://www.mpegla.com/main/programs/avc/Documents/avcweb.pdf), MPEG-LA claims that patent fees can requested by either the OEM, or end-users of software associated with encoding or decoding videos with h264 or h265. As such, conda-forge is interested in understanding what the implications might be for both:

1. conda-forge as an independent entity.
2. The users of the packages distributed by conda-forge.

Statistics publically available on [`anaconda.org`](https://anaconda.org/conda-forge/x264/files) set the number of downloads of the x264 package on the order 100,000 times. It is difficult to estimate how many of these downloads originate from end-users or from automated bots comissioned by conda-forge to compile or test other software packages.

We are interested in better undering 3 broad points:
1. conda-forge's potential financial liability in distributing these algorithms.
2. The end user's potential financial liability in using this software.
3. How the cisco distribution of openh264 can help mitigate the above 2 points.

# conda-forge's liability
Our first question pertains to the liability of conda-forge in distributed these precompiled binary packages that include algorithms claimed by MPEG-LA. Can conda-forge be held liable for its actions related x264 and openh264 even though conda-forge does not generate video or audio content using this software?

For specificity, conda-forge makes use of freely available online resources to:
1. Download the source code of x264/openh264.
2. Compile the source code using bots from Appveyor, Travis CI, CircleCI, Azure Webservices.
3. Test the software. The bots may create small video files to test the correctnenss of the compiled software. Whatever video files created are immediated deleted after the tests.
4. Software is uploaded to anaconda.org for end-users and other bots to download.

If conda-forge can be held liable for such actions, what should conda-forge do to mitigate our legal liability?

# Community's liability
The conda-forge community includes research institutions, private companies, headquartered both in the USA and abroad, and hobbyists. Many community users volunteer their time to help conda-forge in many of its activities.

Therefore, we are interested in relaying to them any unexpected legal consequences that they may incur from downloading and using software from conda-forge.

Could they be held liable for using x264 and openh264 dowloaded from the conda-forge channel on `anaconda.org`?

If the conda-forge community can be held liable for such actions, what should they do to mitigate their legal liability?

# Cisco's distribution of openh264

[Cisco claims](https://www.openh264.org/) that if end-users download the openh264 binaries uploaded by their team, they will pay the patent fees and not pass on those fees to the end user.


1. Can conda-forge download this software and place it in their own archive hosted on an other website (i.e. `anaconda.org`)?
2. If not, can conda forge distribute a script that when executed, would automatically download the binaries from Cisco?



Thank you for your help with these questions.

The team at conda-forge

</details>

editable version: https://hackmd.io/3egc0VI6R7WzazDqneXAzw#

**Comment by scopatz** [comment]
My read on this whole situation is that we have to use the Cisco provided binaries to avoid licensing (which we can't afford) and we have to make it so that `conda install` actually does the download, as per the [binary license](http://www.openh264.org/BINARY_LICENSE.txt).

I actually think that this is pretty straight forward and we don't need to bother the lawyer about it. We might also want to mark all current versions of this package as broken, or something.

**Comment by hmaarrfk** [comment]
@scopatz I tend to agree. I think that would give the best user experience, albeit at the expense of additional complexity at this stage.

It seems that we are OK including x264 and x265, the users will just have to deal with the results right?

**Comment by scopatz** [comment]
I am not sure about those. I haven't looked at them

**Comment by scopatz** [comment]
here we really just need an activate script that downloads and installs the package.
